File: sorting/MergeSort.js
const mergeSort = (arr) => {
    if (arr.length < 2)
        return arr;
 
    const middle = parseInt(arr.length / 2);
    const left   = arr.slice(0, middle);
    const right  = arr.slice(middle, arr.length);
 
    return merge(mergeSort(left), mergeSort(right));
};
 
const merge = (left, right) => {
    const result = [];
 
    while (left.length && right.length) {
        if (left[0] <= right[0]) {
            result.push(left.shift());
        } else {
            result.push(right.shift());
        }
    }
 
    while (left.length)
        result.push(left.shift());
 
    while (right.length)
        result.push(right.shift());
 
    return result;
};

console.log(mergeSort([3,5,6,7,2,4,1,7,4,2]))

File: sorting/MergeSort.rb
#Recursive Merge sort implementation for two sorted lists 

def merge_sort(l, r, a=[])
  # When both lists are empty return the merged array
  if l.empty? && r.empty?
    return a
  end
  # If one list is empty then add the remaining list to the end of the merged array and return that array 
  if l.empty?
    a.concat(r)
    return a
  elsif r.empty?
    a.concat(l)
    return a
  end
  # Compare elements at 0th index
  # Append the lesser integer to the merged array
  # If elements are equal append the 0th index from the right array to the merged array
  l[0] >= r[0] ? a << r.shift : a << l.shift
  # Recursive call to merge_sort
  merge_sort(l,r,a)
end

# Sorted lists
left = [-1,3,7,8,12,45]
right = [-3,-2,5,7,10,11,12,34,56,78]

merge_sort(left,right)